File: scripts/capture_poses.py
#!/usr/bin/env python

import string

import rospy
import rospkg
import rosbag

from sensor_msgs.msg import JointState
from robot_calibration_msgs.msg import CaptureConfig

# Assumes that abb_irb1200_5_90_moveit_config demo.launch is running
# Execute with "rosrun camera_pose_capture capture_poses.py"
# Subscribes to JointState and records msgs into a .bag file
# .bag is saved to the root of the package
# Rough UI loop allows user to dictate what poses are captured
class CapturePoses:
    last_state_ = None # last joint states

    def __init__(self):
        self.last_state_ = CaptureConfig()

        rospy.init_node('capture_calibration_poses')
        rospy.Subscriber("joint_states", JointState, self.stateCb)

        # bag to write data to
        rospack = rospkg.RosPack()
        pkgpath = rospack.get_path('camera_pose_capture')
        bag = rosbag.Bag(pkgpath + '/calibration_poses.bag', 'w')

        # put samples in
        count = 0
        while not rospy.is_shutdown():
            print('Move arm/head to a new sample position:  (type "exit" to quit and save data)')
            resp = raw_input('press <enter> ')
            if string.upper(resp) == 'EXIT':
                break
            else:
                if self.last_state_ == None:
                    print('Cannot save state')
                    continue
                # save a pose
                print(self.last_state_)
                bag.write('/calibration_joint_states', self.last_state_)
                print("Saving pose %d\n" % count)
                count += 1
        bag.close()

    def stateCb(self, msg):
        """ Callback for joint_states messages """
        for joint, position in zip(msg.name, msg.position):
            try:
                idx = self.last_state_.joint_states.name.index(joint)
                self.last_state_.joint_states.position[idx] = position
            except ValueError:
                self.last_state_.joint_states.name.append(joint)
                self.last_state_.joint_states.position.append(position)

if __name__ == '__main__':
    CapturePoses()

File: src/capture_poses.cpp
#include "ros/ros.h"
#include "ros/package.h"
#include "sensor_msgs/JointState.h"
#include "rosbag/bag.h"

#include <iostream>
#include <string>
using namespace std;

/*
 * Helper class used to store robot joint positions and write them to a .bag file
 */
class JointCapture {
public:
	static sensor_msgs::JointState joint_state;
	static void JointCallback(const sensor_msgs::JointState &msg);
	static void WriteToBag(rosbag::Bag *bag);
};

inline sensor_msgs::JointState CreateJointState() {
	sensor_msgs::JointState temp_msg;
	temp_msg.name.push_back("");
	temp_msg.position.push_back(0);
	return temp_msg;
};
sensor_msgs::JointState JointCapture::joint_state = CreateJointState();

void JointCapture::JointCallback(const sensor_msgs::JointState &msg) {
	JointCapture::joint_state = msg;
}

void JointCapture::WriteToBag(rosbag::Bag *bag) {
	bag->write("/joint_states", ros::Time::now(), JointCapture::joint_state);
}

/*
 * Debug function
 * Outputs the name and position of every joint being captured
 */
void PrintJointState() {
	cout << "\n";
	for(int i=0; i<JointCapture::joint_state.name.size(); i++) {
		ROS_INFO("%s: %f", JointCapture::joint_state.name[i].c_str(), JointCapture::joint_state.position[i]);
	}
}

void PrintHelp() {
	ROS_INFO("Press <ENTER> to capture a pose.");
	ROS_INFO("Type \"Exit\" to quit.");
	ROS_INFO("Type \"Help\" to repeat these messages.");
}

/*
 * Assumes that abb_irb1200_5_90_moveit_config demo.launch is running
 * Execute with "rosrun camera_pose_capture capture_poses"
 * Subscribes to JointState and records msgs into a .bag file
 * .bag is saved to the root of the package
 * Rough UI loop allows user to dictate what poses are captured
 */
int main(int argc, char **argv)
{
	ros::init(argc, argv, "capture_poses");
	ros::NodeHandle nh;
	ros::Subscriber joint_sub = nh.subscribe("joint_states", 1, JointCapture::JointCallback);

	rosbag::Bag bag;
	string bag_path = ros::package::getPath("camera_pose_capture");
	bag.open(bag_path + "/calibration_poses_cpp.bag", rosbag::bagmode::Write);

	string input;
	PrintHelp();
	//PrintJointState();
	while(ros::ok())
	{
		cout << "Input: ";
		getline(cin, input);
		for_each(input.begin(), input.end(), [](char & c){
    		c = ::tolower(c);
		});

		if(input.empty())
		{
			ROS_INFO("-- Capturing Pose --");
			ros::spinOnce();
			JointCapture::WriteToBag(&bag);
			//PrintJointState();

		}
		else if(input == "exit")
		{
			ROS_INFO("Saving .bag and quitting");
			break;
		}
		else if(input == "help")
		{
			PrintHelp();
		}
		else
		{
			ROS_INFO("Oh ffs what?");
		}
	}

	bag.close();
	return 0;
}